### ISSUE TITLE
NuGet CLI Bulk Unlister Tool

### ISSUE THREAD

**Issue by jcjiang** [NuGet/Home#9897]
## **Problem Background**

Within the NuGet CLI, authors can choose either to [unlist](https://docs.microsoft.com/en-us/nuget/nuget-org/policies/deleting-packages#unlisting-a-package) or [deprecate](https://docs.microsoft.com/en-us/nuget/nuget-org/deprecate-packages) particular versions of their published packages to limit public access. There does not currently exist an efficient way to deprecate or unlist multiple or all versions of a package. This presents a process inefficiency that can be potentially significant for customers who must often unlist versions of their package due to security or dependency concerns.  In addition, having to manually unlist package versions can result in human error and lack of thoroughness.

## **Who are the customers**

Specific use cases are:

1. Customers who want to unlist all versions of their published package efficiently.
2. Customers who want to unlist most of the versions, with individual exceptions.
3. Customers who want to unlist a certain range of versions (i.e. 3.x and lower)
4. Customers who want to unlist versions with names that include a particular regular expression pattern. 

## Goals and Success Metrics

**Process Efficiency**

- Provide customers a faster and easier process for unlisting multiple or all versions of a package

**Control**

- Offer customers more control over the unlisting of their package versions.

**Security**

- Drive broader usage of the unlisting functionality to eliminate package vulnerabilities, bugs, and dependency errors across versions

## **Non-Goals**

**[NuGet.org](http://nuget.org) and Visual Studio**

This solution is for the NuGet CLI only. There will be no discussion surrounding UI on [NuGet.org](http://nuget.org) and the Visual Studio version of NuGet

**Deprecation**

This solution is for the unlisting functionality only. There will be no work done on bulk-deprecation.

## **Solution**

The solution will come in a form on a downloadable CLI tool. The first version of this tool will be straightforward and follows the following steps:

1. User invokes "unlist" command 
2. Existing versions are checked against the criteria for unlisting provided by the user
3. List of versions to be unlisted is printed to console. User is prompted to confirm.
    1. If user confirms, the tool proceeds to unlist all versions agreed upon
    2. If user does not confirm, the tool terminates and allows the user to revise their query
4. Once the process is complete, a list of the unlisted files will be printed to console

## **What:**

[Very simplified architecture](https://www.notion.so/Architecture-for-Bulk-Unlister-f7b062984b07440ebade5a0693d6549e) can be found here

[Specific CLI command examples to come]

**Comment by michael-hawker** [comment]
@agr this was closed as completed, but there's no mention of bulk unlist [in the docs](https://learn.microsoft.com/nuget/nuget-org/policies/deleting-packages). How do we do this? It seems as part of the CLI docs `<packageVersion>` is listed as optional, but the behavior of not specifying it isn't defined: https://learn.microsoft.com/nuget/reference/cli-reference/cli-ref-delete

I'd really just like to see the same UI that deprecation has to 'Select all versions' for listing too:
![image](https://github.com/user-attachments/assets/880e6788-e57a-49ca-95f3-5b391bae56c8)

**Comment by lilith** [comment]
This does not seem to be completed. Please reopen?
